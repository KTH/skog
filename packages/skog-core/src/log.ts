import { getFields } from "./asyncContext";

export const LEVEL_NUMBERS = {
  fatal: 60,
  error: 50,
  warn: 40,
  info: 30,
  debug: 20,
  trace: 10,
};

export type Levels = keyof typeof LEVEL_NUMBERS;

export interface LogFunction {
  (arg1: Error | Record<string, any>, message: string): void;

  (message: string): void;
}

export interface Logger {
  fatal: LogFunction;
  error: LogFunction;
  warn: LogFunction;
  info: LogFunction;
  debug: LogFunction;
  trace: LogFunction;
}

function print(level: Levels, arg1: string | Error | any, arg2?: unknown) {
  const fields = getFields();

  if (fields) {
    if (typeof arg1 === "string") {
      console.log("[%s] %s %o", level, arg1, fields);
    } else if (typeof arg2 === "string") {
      console.log("[%s] %s %o", level, arg2, { ...fields, ...arg1 });
    }
  } else {
    if (typeof arg1 === "string") {
      console.log("[%s] %s", level, arg1);
    } else if (typeof arg2 === "string") {
      console.log("[%s] %s %o", level, arg2, arg1);
    }
  }
}

export const consoleLog: Logger = {
  fatal: (arg1, arg2?) => print("fatal", arg1, arg2),
  error: (arg1, arg2?) => print("error", arg1, arg2),
  warn: (arg1, arg2?) => print("warn", arg1, arg2),
  info: (arg1, arg2?) => print("info", arg1, arg2),
  debug: (arg1, arg2?) => print("debug", arg1, arg2),
  trace: (arg1, arg2?) => print("trace", arg1, arg2),
};
