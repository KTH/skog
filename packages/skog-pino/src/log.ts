import { Logger as SkogLogger, getFields, Levels } from "skog-core";
import pino, {
  LoggerOptions as PinoLoggerOptions,
  DestinationStream,
  Logger as PinoLogger,
} from "pino";

let instance: PinoLogger;

export function initializeLogger(
  optionsOrStream?: PinoLoggerOptions | DestinationStream
): void;
export function initializeLogger(
  options: PinoLoggerOptions,
  stream: DestinationStream
): void;
export function initializeLogger(
  optionsOrStream?: PinoLoggerOptions | DestinationStream,
  stream?: DestinationStream
) {
  if (!stream) {
    instance = pino(optionsOrStream);
  } else {
    instance = pino(optionsOrStream as PinoLoggerOptions, stream);
  }
}

function mergeObject(obj: Error | object | string) {
  const fields = getFields();

  if (obj instanceof Error) {
    return {
      msg: "",
      err: obj,
      ...fields,
    };
  } else if (typeof obj === "string") {
    return {
      msg: obj,
      ...fields,
    };
  } else {
    return {
      msg: "",
      ...fields,
      ...obj,
    };
  }
}

function print(level: Levels, arg1: string | Error | any, arg2: unknown) {
  if (!instance) {
    throw new Error(
      "It is not possible to log stuff before instantiating. Use `initializeLogger` first"
    );
  }

  const loggerFunction = instance[level].bind(instance);
  const obj = mergeObject(arg1);

  if (typeof arg2 === "string") {
    obj.msg = arg2;
  }

  loggerFunction(obj);
}

export { SkogLogger as Logger };
export const log: SkogLogger = {
  trace: (arg1, arg2?) => print("trace", arg1, arg2),
  debug: (arg1, arg2?) => print("debug", arg1, arg2),
  info: (arg1, arg2?) => print("info", arg1, arg2),
  warn: (arg1, arg2?) => print("warn", arg1, arg2),
  error: (arg1, arg2?) => print("error", arg1, arg2),
  fatal: (arg1, arg2?) => print("fatal", arg1, arg2),
};
