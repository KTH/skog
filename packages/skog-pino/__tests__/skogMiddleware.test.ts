import log, {
  skogMiddleware,
  runWithSkogContext,
  initializeLogger,
} from "../src/index";
import express from "express";
import { setTimeout } from "timers/promises";
import http from "http";

// Define a HTTP server and client
const app = express();
app.use(skogMiddleware);
app.get("/", async (req, res) => {
  for (let i = 0; i < 10; i++) {
    await setTimeout(Math.random() * 100);
    log.info(`${req.query.name} - Hello`);
  }

  res.send(req.query.name);
});

async function sendRequest(name: string) {
  return new Promise<void>((resolve) => {
    http.get(`http://localhost:3001?name=${name}`, (res) => {
      resolve();
    });
  });
}

describe("skogMiddleware", () => {
  let server: http.Server;
  let result: string[] = [];

  beforeAll(() => {
    server = app.listen(3001);
    jest.useFakeTimers("modern").setSystemTime(new Date(1234));
  });

  beforeEach(() => {
    result = [];
    initializeLogger(
      {},
      {
        write(msg) {
          result.push(msg);
        },
      }
    );
  });

  test("Ensure that `console.log` is called 10 times with the same arguments", async () => {
    const names = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J"];
    await Promise.all(names.map((name) => sendRequest(name)));

    // Ensure all calls that contain "A - Hello" have the same content
    const logsA = result.filter((r) => r.includes("A - Hello"));
    expect(logsA.filter((l) => l === logsA[0])).toEqual(logsA);
  });

  afterAll(() => {
    server.close();
  });
});

describe("log", () => {
  let result: string[] = [];

  beforeAll(() => {
    jest.useFakeTimers("modern").setSystemTime(new Date(1234));
  });

  beforeEach(() => {
    result = [];
    initializeLogger(
      { base: undefined },
      {
        write(msg) {
          result.push(msg);
        },
      }
    );
  });

  test("Ensure log methods renders Error objects correctly", () => {
    try {
      throw new Error("Example error");
    } catch (err) {
      if (err instanceof Error) {
        log.info(err);
        log.info(err, "Custom message");
      }
    }

    expect(result.length).toBe(2);
    expect(result[0]).toMatchSnapshot();
    expect(result[1]).toMatchSnapshot();
  });

  test("Ensure log methods renders Error objects within a context", () => {
    runWithSkogContext({ id: 1 }, () => {
      try {
        throw new Error("Example error");
      } catch (err) {
        if (err instanceof Error) {
          log.info(err);
          log.info(err, "Custom message");
        }
      }
    });

    expect(result.length).toBe(2);
    expect(result[0]).toMatchSnapshot();
    expect(result[1]).toMatchSnapshot();
  });
});
