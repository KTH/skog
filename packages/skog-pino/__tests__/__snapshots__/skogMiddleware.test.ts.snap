// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log Ensure log methods renders Error objects correctly 1`] = `
"{\\"level\\":30,\\"time\\":1234,\\"msg\\":\\"\\",\\"err\\":{\\"type\\":\\"Error\\",\\"message\\":\\"Example error\\",\\"stack\\":\\"Error: Example error\\\\n    at Object.<anonymous> (/Volumes/Data/kth/e-larande/skog/packages/skog-pino/__tests__/skogMiddleware.test.ts:86:13)\\\\n    at Promise.then.completed (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/utils.js:391:28)\\\\n    at new Promise (<anonymous>)\\\\n    at callAsyncCircusFn (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/utils.js:316:10)\\\\n    at _callCircusTest (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:218:40)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\\\n    at _runTest (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:155:3)\\\\n    at _runTestsForDescribeBlock (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:66:9)\\\\n    at _runTestsForDescribeBlock (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:60:9)\\\\n    at run (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:25:3)\\"},\\"msg\\":\\"Example error\\"}
"
`;

exports[`log Ensure log methods renders Error objects correctly 2`] = `
"{\\"level\\":30,\\"time\\":1234,\\"msg\\":\\"Custom message\\",\\"err\\":{\\"type\\":\\"Error\\",\\"message\\":\\"Example error\\",\\"stack\\":\\"Error: Example error\\\\n    at Object.<anonymous> (/Volumes/Data/kth/e-larande/skog/packages/skog-pino/__tests__/skogMiddleware.test.ts:86:13)\\\\n    at Promise.then.completed (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/utils.js:391:28)\\\\n    at new Promise (<anonymous>)\\\\n    at callAsyncCircusFn (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/utils.js:316:10)\\\\n    at _callCircusTest (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:218:40)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\\\n    at _runTest (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:155:3)\\\\n    at _runTestsForDescribeBlock (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:66:9)\\\\n    at _runTestsForDescribeBlock (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:60:9)\\\\n    at run (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/run.js:25:3)\\"},\\"msg\\":\\"Example error\\"}
"
`;

exports[`log Ensure log methods renders Error objects within a context 1`] = `
"{\\"level\\":30,\\"time\\":1234,\\"msg\\":\\"\\",\\"err\\":{\\"type\\":\\"Error\\",\\"message\\":\\"Example error\\",\\"stack\\":\\"Error: Example error\\\\n    at /Volumes/Data/kth/e-larande/skog/packages/skog-pino/__tests__/skogMiddleware.test.ts:102:15\\\\n    at /Volumes/Data/kth/e-larande/skog/packages/skog-core/src/asyncContext.ts:58:12\\\\n    at /Volumes/Data/kth/e-larande/skog/packages/skog-core/src/asyncContext.ts:45:64\\\\n    at AsyncLocalStorage.run (node:async_hooks:320:14)\\\\n    at /Volumes/Data/kth/e-larande/skog/packages/skog-core/src/asyncContext.ts:45:27\\\\n    at runWithSkogContext (/Volumes/Data/kth/e-larande/skog/packages/skog-core/src/asyncContext.ts:59:5)\\\\n    at Object.<anonymous> (/Volumes/Data/kth/e-larande/skog/packages/skog-pino/__tests__/skogMiddleware.test.ts:100:23)\\\\n    at Promise.then.completed (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/utils.js:391:28)\\\\n    at new Promise (<anonymous>)\\\\n    at callAsyncCircusFn (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/utils.js:316:10)\\"},\\"id\\":1,\\"msg\\":\\"Example error\\"}
"
`;

exports[`log Ensure log methods renders Error objects within a context 2`] = `
"{\\"level\\":30,\\"time\\":1234,\\"msg\\":\\"Custom message\\",\\"err\\":{\\"type\\":\\"Error\\",\\"message\\":\\"Example error\\",\\"stack\\":\\"Error: Example error\\\\n    at /Volumes/Data/kth/e-larande/skog/packages/skog-pino/__tests__/skogMiddleware.test.ts:102:15\\\\n    at /Volumes/Data/kth/e-larande/skog/packages/skog-core/src/asyncContext.ts:58:12\\\\n    at /Volumes/Data/kth/e-larande/skog/packages/skog-core/src/asyncContext.ts:45:64\\\\n    at AsyncLocalStorage.run (node:async_hooks:320:14)\\\\n    at /Volumes/Data/kth/e-larande/skog/packages/skog-core/src/asyncContext.ts:45:27\\\\n    at runWithSkogContext (/Volumes/Data/kth/e-larande/skog/packages/skog-core/src/asyncContext.ts:59:5)\\\\n    at Object.<anonymous> (/Volumes/Data/kth/e-larande/skog/packages/skog-pino/__tests__/skogMiddleware.test.ts:100:23)\\\\n    at Promise.then.completed (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/utils.js:391:28)\\\\n    at new Promise (<anonymous>)\\\\n    at callAsyncCircusFn (/Volumes/Data/kth/e-larande/skog/node_modules/jest-circus/build/utils.js:316:10)\\"},\\"id\\":1,\\"msg\\":\\"Example error\\"}
"
`;
